blackbird_python/blackbird/listener.py

```diff
@@ -116,84 +118,54 @@ def __str__(self):
-
-        self.active_modes = set()
-        """set[int]: A set of non-negative integers specifying the modes the program manipulates."""
-
-        self.var = {}
-        """dict[str->[int, float, complex, str, bool, numpy.ndarray]]: Mapping from the
-        variable names in the Blackbird script to their declared values."""
-
-        self.target = None
-        """dict[str->[str, dict]]: Contains information regarding the target device of the quantum
-        program (i.e., the target device the Blackbird script is compiled for).
+    On initialization, an empty :class:`~.BlackbirdProgram` object is initialized.
+    Over the course of the parsing, the program details are filled in according to the
+    parsed script.
 
-        Important keys include:
-
-        * ``'name'`` (str): the name of the device the Blackbird script requests to be run on
-        * ``'options'`` (dict): a dictionary of keyword arguments for the target device
-        """
-
-        self.queue = []
-        """list[dict]: List of operations to apply to the device, in temporal order.
-        Each operation is contained as a dictionary, with the following keys:
+    Once parsing is complete, the :class:`~.BlackbirdProgram` object can be returned
+    via the :attr:`program` attribute.
+    """
 
-        * ``'op'`` (str): the name of the operation
-        * ``'args'`` (list): a list of positional arguments for the operation
-        * ``'kwargs'`` (dict): a dictionary of keyword arguments for the operation
-        * ``'modes'`` (list[int]): modes the operation applies to
+    def __init__(self):
+        self._program = BlackbirdProgram()
 
-        Note that, depending on the operation, both ``'args'`` and ``'kwargs'``
-        might be empty.
-        """
+    @property
+    def program(self):
+        """Returns the parsed blackbird program"""
+        return self._program
 
     def exitDeclarename(self, ctx: blackbirdParser.DeclarenameContext):
         """Run after exiting program name metadata.
 
         Args:
             ctx: DeclarenameContext
         """
-        self.name = ctx.programname().getText()
+        self._program._name = ctx.programname().getText()
 
     def exitVersion(self, ctx: blackbirdParser.VersionContext):
         """Run after exiting version metadata.
 
         Args:
             ctx: VersionContext
         """
-        self.version = ctx.versionnumber().getText()
+        self._program._version = ctx.versionnumber().getText()
 
     def exitTarget(self, ctx: blackbirdParser.TargetContext):
         """Run after exiting target metadata.
 
         Args:
             ctx: TargetContext
         """
-        self.target = {"name": ctx.device().getText()}
+        self._program._target["name"] = ctx.device().getText()
 
-        args = []
         kwargs = {}
 
         if ctx.arguments():
-            args, kwargs = _get_arguments(ctx.arguments())
+            _, kwargs = _get_arguments(ctx.arguments())
```

Comment:
Any want/need to warn that `args` get ignored?

Reply:
Good idea. I'll raise a warning if the user tries to put positional args here.